### DISCUSSION TITLE
Framework proposals

### DISCUSSION THREAD

**Discussion by Kiarancole** [uprm-inso4101-2024-2025-s1/semester-project-campus-resource-and-event-management#22] [Ideas]
Space for proposing web app frameworks we can use

**Comment by Ojani** [comment]
# Svelte Frontend Framework Proposal

## Svelte vs Sveltekit
Svelte and Sveltekit are both made by the same people, but Svelte is specifically for frontend and Sveltekit is for backend. They obviously work well together since they were made for that, but here I want to mainly make the case for using Svelte specifically. The good thing about using Sveltekit, however, is that it is a JavaScript Backend framework, which means that if Sveltekit is used, the frontend and the backed will be written in the same language, JavaScript.

## Why Use a Frontend Framework
Most modern frontend frameworks have a component-based approach in which you can build reusable components which can have different properties passed on to them to make them behave how you want. For example, maybe you want to have a button throughout the web app. Well, you can make a component for that button which contains the HTML and CSS for the button and reuse the component anywhere just by importing the component. This is nice because you avoid having to rewrite HTML and since components are always contained in their own separate file, it makes the code a lot more readable and less overwhelming. Finally, if any change has to be made, it is reflected in all places where the component is used. Another really good reason to use a framework is reactivity; this means that when a variable is changed, it automatically updates all the parts of the UI that rely on that variable, including any nested components.

## Example pipeline for creating UI components
Creating a component can be as easy as having someone design it in Figma, exporting the HTML and CSS directly from Figma into a dedicated file for that component, adding any javascript necessary for it to work, and importing the component wherever it is needed. This could easily be split into multiple tasks where one would be to create the Figma design and export it, and the other would be to add any additional functionality to the component using JavaScript. 

Additionally, before a component is created, a placeholder component can be made so that when the design is finished, it appears wherever the placeholder was used. For example, we could have a placeholder component called "Button" which will be used wherever a button is needed and it will only be composed of an unstyled HTML button. Later on, someone fulfills the task of creating the design and the Button component is updated, automatically updating all the places in which it was used. This is especially useful because it means that certain tasks won't be held back by having to wait for parts of the UI to be designed and it means that we will never have to edit every single page whenever a small redesign is made.

## Why Use Svelte Over Other Frameworks
- It is one of the easiest frontend frameworks to learn
- It has a very intuitive syntax
- It has an amazing official tutorial which teaches you everything you need to know and can be completed in a weekend
- Unlike other frameworks such as React, it has many tools that come out of the box, which avoids slowing the development process because of having to find different libraries, such as for state management.
- Any other JavaScript library works perfectly fine with Svelte without having to do any extra configuration
- There are many code examples (REPLs) available online which can show you how to do something in an interactive environment
- components can be written in their online code editor for testing things
- It includes transitions and animations out of the box, which allows for the easy creation of a fluid user experience
- It is compiled into a static output which only includes what it needs to function. This means that the web app will not hog up memory (like React [I'm sorry React]) in order to keep track of UI updates and that sorta thing.
- In the 2024 Stack Overflow Developer Survey, it was voted the 2nd most admired component framework by its users with a 72.8% admiration rate (React at 62.2%, VueJS at 60.2%, and Angular at 23.1% for reference), only coming second to Phoenix.

## How to Get Started
To learn to use the framework is extremely easy and doesn't require you to download anything, all you have to do is go to [their website](https://learn.svelte.dev/tutorial/welcome-to-svelte) to get started. If we're only using Svelte and not Sveltekit, only part 1 and 2 of the tutorial are needed and can be learned in a weekend.

## Reactive Button Click Counter Example Component
### Svelte
```html
<script>
  let count = 0;
</script>

<button on:click={() => count++}>
  clicks: {count}
</button>
```

### React
```javascript
import { useState } from "react";

function Counter() {
  const [count, setCount] = useState(0);
  return (
    <button onClick={() => setCount(count + 1)}>
      clicks: {count}
    </button>
  );
}

export default Counter;
```

## Final notes
If we decide to use Svelte, we should stick to Svelte 4. Svelte 5 is nearly done and can be used already, but things are still subject to change and to avoid confusion we should all be on the same page and use Svelte 4. Svelte 5 will be fully compatible with Svelte 4 and all the old code will work, but still, lets stick to 4 ❤️.

**Reply by vivianaramos6** [reply]
I support this

**Comment by Adrianortiz-PR** [comment]
# Python Flask Backend Framework Proposal

## Flask Overview
Flask is a micro-framework for Python that's widely used for building web applications and APIs. It’s known for being lightweight, flexible, and easy to use. Flask allows developers to get started quickly with minimal setup while also being powerful enough to scale up for larger applications.

By using Flask for the backend, the app would benefit from Python's rich ecosystem, including numerous libraries for data manipulation, AI, machine learning, and database interaction. Python is well-known for its simplicity and readability, making Flask an ideal choice for building maintainable, scalable applications.

## Why Use Flask Over Other Backend Frameworks

### Lightweight and Simple
Flask is known for being minimalistic and does not impose too many dependencies or architectural decisions. Developers have the flexibility to use only the tools they need, and they can add extensions as the project grows.

### Python Ecosystem
Python is one of the most popular programming languages, making Flask a great option for back-end development, especially if the team is already familiar with Python. Python's ecosystem is rich with libraries and tools that can easily integrate into the app, from AI and data science tools to web scraping and more.

### Customizability
Flask doesn't force developers into any specific project layout, allowing flexibility. It’s especially good for smaller applications or those that need to be tailored to specific project requirements. You are free to choose your database, frontend framework (like Svelte), and additional tools like Redis or Celery for task queues.

### RESTful API Design
Flask makes it easy to build REST APIs, which would be ideal for serving data to the Svelte frontend. By using Flask, we can easily build clean, well-documented APIs for various app features. Flask is particularly good at handling asynchronous API requests and responses, making it a great choice for dynamic web apps.

### Flask Extensions
Flask has a wide range of extensions available that can add functionalities like user authentication (Flask-Login), form handling (Flask-WTF), database management (Flask-SQLAlchemy), and more. These extensions can be added when necessary, keeping the core app lightweight while ensuring it has all the necessary features.

## Example Workflow for Flask Development

1. **Setting Up Routes**: Flask uses Python decorators to define routes. This allows us to create endpoints for serving the necessary data to the frontend (Svelte).
    ```python
    from flask import Flask, jsonify

    app = Flask(__name__)

    @app.route('/api/data', methods=['GET'])
    def get_data():
        data = {"message": "Hello from Flask!"}
        return jsonify(data)

    if __name__ == "__main__":
        app.run(debug=True)
    ```

2. **Database Integration**: Flask can be paired with SQLAlchemy or any other ORM to interact with a database. This allows the app to store and manage data seamlessly.
    ```python
    from flask_sqlalchemy import SQLAlchemy
    db = SQLAlchemy(app)
    ```

3. **Serving Data to Svelte**: The Flask backend will serve data to the Svelte frontend through API endpoints. This approach separates concerns between the frontend and backend and ensures the app is scalable and easy to maintain.

## Flask vs. Other Backend Options

- **Django**: While Django offers more built-in functionality, it's a heavier framework than Flask and is often considered overkill for small to medium projects. Flask’s simplicity offers better flexibility and faster development times, especially when paired with a frontend like Svelte.
  
- **Node.js (SvelteKit)**: SvelteKit is an excellent option if you want both your frontend and backend to be in JavaScript. However, Python offers distinct advantages for data-heavy applications and scientific computing, which may be beneficial depending on your project.





## Final Notes
Flask is well-suited for the backend of this project due to its lightweight nature, flexibility, and seamless integration with Python's rich ecosystem. It allows us to easily build RESTful APIs and pair them with a dynamic Svelte frontend.

By choosing Flask, we ensure that the backend is scalable, maintainable, and adaptable for future feature expansions. Additionally, Flask’s learning curve is small, meaning new developers can quickly become productive. Let’s stick to Flask 2.x to maintain consistency and stability while taking advantage of its recent updates. 

This combination of Flask for the backend and Svelte for the frontend provides a solid foundation for building a fast, responsive web application that can grow with the project's needs.
